### ISSUE TITLE
Proxies for lazy loaded dependencies are not written to file?

### ISSUE THREAD

**Issue by holtkamp** [PHP-DI/PHP-DI#411]
When:

- configuring objects to be lazily loaded using the ObjectDefinitionHelper as suggested in the [docs](http://php-di.org/doc/lazy-injection.html#php-configuration-file):
```php
//definitions.php
array(
    Class1::class => object()->lazy(),
    Class2::class => object()->lazy(),
    ClassN::class => object()->lazy(),
)
```
- storing the Definitions in a file and adding them like ```$builder->addDefinitions('definitions.php');```
- using the ```ocramius\proxy-manager```
- configuring the DI builder to write the proxies to a file: ```writeProxiesToFile(true, '/folderName')```

Then:
- no files are actually generated in the indicated folder
- performance drops significantly. 

When switching back to non-lazy configured objects, performance is back to normal again. After browsing around in the source code, it seems a ```FileWriterGeneratorStrategy``` should also be set explicitly, since [it defaults](https://github.com/Ocramius/ProxyManager/blob/6e15877fc03cc211ce798d497a21208a44f7b726/src/ProxyManager/Configuration.php#L146-L150) to an ```EvaluatingGeneratorStrategy``` which [removes the generated files afterwards](https://github.com/Ocramius/ProxyManager/blob/6e15877fc03cc211ce798d497a21208a44f7b726/src/ProxyManager/GeneratorStrategy/EvaluatingGeneratorStrategy.php#L62)

Adding one line to the ```ProxyFactory``` https://github.com/PHP-DI/PHP-DI/blob/c6c932e682c9e7bd009405913acfc05f4693ed60/src/DI/Proxy/ProxyFactory.php#L72-L74 resolves this:
```
$config->setGeneratorStrategy(new FileWriterGeneratorStrategy(new \ProxyManager\FileLocator\FileLocator($this->proxyDirectory)));
```

Also opened https://github.com/PHP-DI/PHP-DI/pull/412

**Comment by mnapoli** [comment]
Oh wow thanks. This is not great indeed, but I think this is because of a BC break in ProxyManager 2.0 (https://github.com/Ocramius/ProxyManager/releases/tag/2.0.0): https://github.com/Ocramius/ProxyManager/issues/105

The "officially recommended" version is ~1.0 (through docs and https://github.com/PHP-DI/PHP-DI/blob/master/composer.json#L44). `suggests` is showing its limits here. However it's of course great that we add support for 1.0 and 2.0, it was mentioned in https://github.com/PHP-DI/PHP-DI/issues/359#issuecomment-202996701 but I never got around doing it. If you could update your PR to keep the `else { $config->setGeneratorStrategy(new EvaluatingGeneratorStrategy());` that way the factory would work with both versions?

If you have time you could also update composer.json and the docs, but if you don't that's not a problem I'll probably have some time later today for that.

**Comment by holtkamp** [comment]
> If you could update your PR to keep the else { $config->setGeneratorStrategy(new EvaluatingGeneratorStrategy()); that way the factory would work with both versions?

Done!

> If you have time you could also update composer.json and the docs, but if you don't that's not a problem I'll probably have some time later today for that.

Don't know whether that is still required, both versions are supported now right ? 😄

**Comment by mnapoli** [comment]
Fixed in #412
